File: src/ee3305_py/ee3305_py/behavior.py
from math import hypot

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data, qos_profile_services_default
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Odometry, Path


class Behavior(Node):

    def __init__(self, node_name="behavior"):
        # Node Constructor =============================================================
        super().__init__(node_name)

        # Parameters: Declare
        self.declare_parameter("frequency", float(10))
        self.declare_parameter("plan_frequency", float(2))

        # Parameters: Get Values
        self.frequency_ = self.get_parameter("frequency").value
        self.plan_frequency_ = self.get_parameter("plan_frequency").value

        # Handles: Topic Subscribers
        # !TODO: Goal pose subscriber
        self.sub_goal_pose_ = self.create_subscription(
            PoseStamped,               # Message type
            "goal_pose",               # Topic name
            self.callbackSubGoalPose_, # Callback function
            10                         # QoS
        )

        # !TODO: Odometry subscriber
        self.sub_odom_ = self.create_subscription(
            Odometry,
            "odom",
            self.callbackSubOdom_,
            10
        )

        # Handles: Topic Publishers
        # !TODO: Path request publisher
        self.pub_path_request_ = self.create_subscription(
            Path,
            "path_request",
            10
        )

        # Handles: Timers
        self.timer = self.create_timer(1.0 / self.frequency_, self.callbackTimer_)

        self.timer_plan_ = self.create_timer(
            1.0 / self.plan_frequency_, self.callbackTimerPlan_
        )

        # Other Instance Variables
        self.received_goal_coords_ = False
        self.received_rbt_coords_ = False
        self.goal_reached_ = True

    # Callbacks =============================================================

    # Goal pose subscriber callback.
    def callbackSubGoalPose_(self, msg: PoseStamped):
        self.received_goal_coords_ = True
        self.goal_reached_ = (
            True  # "cancel" the goal and trigger the if condition in timer.
        )

        # !TODO: Copy to goal_x_, goal_y_.
        self.goal_x_ = msg.pose.position.x
        self.goal_y_ = msg.pose.position.y

        self.get_logger().info(
            f"Received New Goal @ ({self.goal_x_:7.3f}, {self.goal_y_:7.3f})."
        )

    # Odometry subscriber callback.
    def callbackSubOdom_(self, msg: Odometry):
        self.received_rbt_coords_ = True

        # !TODO: Copy to rbt_x_, rbt_y_.
        self.rbt_x_ = msg.pose.pose.position.x
        self.rbt_y_ = msg.pose.pose.position.y


    # Callback for timer.
    # Normally the decisions of the robot system are made here, and this callback is dramatically simplified.
    # The callback contains some example code for waypoint detection.
    def callbackTimer_(self):
        if not self.received_rbt_coords_ or not self.received_goal_coords_:
            return  # silently return if none of the coords are received from the subscribers.

        dx = self.goal_x_ - self.rbt_x_
        dy = self.goal_y_ - self.rbt_y_

        goal_is_close = hypot(dx, dy) < 0.1

        if goal_is_close and not self.goal_reached_:
            self.get_logger().info(
                f"Reached Goal @ ({self.goal_x_:7.3f}, {self.goal_y_:7.3f})."
            )
            self.goal_reached_ = True
        elif not goal_is_close and self.goal_reached_:
            self.get_logger().info(
                f"Going to Goal @ ({self.goal_x_:7.3f}, {self.goal_y_:7.3f})."
            )
            self.goal_reached_ = False

    # Callback for publishing path requests between clicked_point (goal) and robot position.
    # Normally path requests are implemented with ROS2 service, and the service is called in the main timer.
    # To keep things simple for this course, we use only ROS2 tpics.
    def callbackTimerPlan_(self):
        if not self.received_goal_coords_ or not self.received_rbt_coords_:
            return  # silently return if none of the coords are received from the subscribers

        # Create a new message for publishing
        msg_path_request = Path()
        msg_path_request.header.stamp = self.get_clock().now().to_msg()
        msg_path_request.header.frame_id = "map"

        # !TODO: write the robot coordinates
        rbt_pose = PoseStamped()
        rbt_pose.pose.position.x = self.rbt_x_
        rbt_pose.pose.position.y = self.rbt_y_

        # !TODO: write the goal coordinates
        goal_pose = PoseStamped()
        goal_pose.pose.position.x = self.goal_x_
        goal_pose.pose.position.y = self.goal_y_

        # !TODO: fill up the array containing the robot coordinates at [0] and goal coordinates at [1]
        msg_path_request.poses.append(rbt_pose)
        msg_path_request.poses.append(goal_pose)

        # publish the message
        self.get_logger().info(
            f"Sending Path Planning Request from Rbt @ ({self.rbt_x_:7.3f}, {self.rbt_y_:7.3f}) to Goal @ ({self.goal_x_:7.3f}, {self.goal_y_:7.3f})"
        )
        self.pub_path_request_.publish(msg_path_request)


# Main Boiler Plate =============================================================
def main(args=None):
    rclpy.init(args=args)
    rclpy.spin(Behavior())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/ee3305_cpp/src/behavior.cpp
#include <queue>
#include <iostream>
#include <iomanip>
#include <memory>
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "nav_msgs/msg/path.hpp"

using namespace std::chrono_literals; // required for using the chrono literals such as "ms" or "s"

namespace ee3305
{
    /** Publishes the raw occupancy grid data, and the global costmap that has inflation cost values. */
    class Behavior : public rclcpp::Node
    {
    private:
        // Node Instance Properties =============================================================

        // Handles
        rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_goal_pose_;
        rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
        rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr pub_path_request_; // should use ROS services.
        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::TimerBase::SharedPtr timer_plan_;

        // Data from Subscriptions
        double goal_x_; // read only; written only by subscriber(s)
        double goal_y_; // read only; written only by subscriber(s)
        double rbt_x_;  // read only; written only by subscriber(s)
        double rbt_y_;  // read only; written only by subscriber(s)

        // Parameters
        double frequency_;      // read only; written only in constructor
        double plan_frequency_; // read only; written only in constructor

        // Other Instance Variables
        bool received_goal_coords_;
        bool received_rbt_coords_;
        bool goal_reached_;

    public:
        explicit Behavior(std::string node_name = "behavior")
            : Node(node_name)
        {
            // Node Constructor =============================================================

            // Parameters: Declare (remember to get after declare)
            this->declare_parameter<double>("frequency", 10);
            this->declare_parameter<double>("plan_frequency", 2);

            // Parameters: Get values (remember to declare first)
            this->frequency_ = this->get_parameter("frequency").get_value<double>();
            this->plan_frequency_ = this->get_parameter("plan_frequency").get_value<double>();

            // Handles: Topic Subscribers
            // !TODO: Goal pose subscriber

            // !TODO: Odometry subscriber

            // Handles: Topic Publishers
            // !TODO: Path request publisher

            // Handles: Timers
            this->timer_ = this->create_wall_timer(
                1.0s / this->frequency_, std::bind(&Behavior::callbackTimer_, this));

            this->timer_plan_ = this->create_wall_timer(
                1.0s / this->plan_frequency_, std::bind(&Behavior::callbackTimerPlan_, this));

            // Other Instance Variables
            this->received_goal_coords_ = false;
            this->received_rbt_coords_ = false;
            this->goal_reached_ = true;
        }

    private:
        // Callbacks =============================================================

        // Goal pose subscriber callback.
        void callbackSubGoalPose_(geometry_msgs::msg::PoseStamped::SharedPtr msg)
        {
            this->received_goal_coords_ = true;
            this->goal_reached_ = true; // "cancel" the goal and trigger the if condition in timer.

            // !TODO: Copy to goal_x_, goal_y_.
            this->goal_x_ = msg->pose.position.x;

            RCLCPP_INFO(this->get_logger(),
                        "Received New Goal @ (%7.3f, %7.3f).",
                        this->goal_x_, this->goal_y_);
        }

        // Odometry subscriber callback.
        void callbackSubOdom_(nav_msgs::msg::Odometry::SharedPtr msg)
        {
            this->received_rbt_coords_ = true;

            // !TODO: Copy to rbt_x_, rbt_y_.
            this->rbt_x_ = msg->pose.pose.orientation.y;
        }

        // Callback for timer.
        // Normally the decisions of the robot system are made here, and this callback is dramatically simplified.
        // The callback contains some example code for waypoint detection.
        void callbackTimer_()
        {
            if (!this->received_goal_coords_ || !this->received_rbt_coords_)
                return; // silently return if none of the coords are received from the subscribers.

            double dx = this->goal_x_ - this->rbt_x_;
            double dy = this->goal_y_ - this->rbt_y_;

            bool goal_is_close = std::hypot(dx, dy) < 0.1;

            if (goal_is_close && !this->goal_reached_)
            { // reached the goal pose
                RCLCPP_INFO(this->get_logger(),
                            "Reached Goal @ (%7.3f, %7.3f).",
                            this->goal_x_, this->goal_y_);
                this->goal_reached_ = true;
            }
            else if (!goal_is_close && this->goal_reached_)
            {
                RCLCPP_INFO(this->get_logger(),
                            "Going to Goal @ (%7.3f, %7.3f).",
                            this->goal_x_, this->goal_y_);
                this->goal_reached_ = false;
            }
        }

        // Callback for publishing path requests between clicked_point (goal) and robot position.
        // Normally path requests are implented with ROS2 service, and the service is called in the main timer.
        // To keep things simple for this course, we use only ROS2 tpics.
        void callbackTimerPlan_()
        {
            if (!this->received_goal_coords_ || !this->received_rbt_coords_)
                return; // silently return if none of the coords are received from the subscribers

            // Create a new message for publishing
            nav_msgs::msg::Path msg_path_request;
            msg_path_request.header.stamp = this->get_clock()->now();
            msg_path_request.header.frame_id = "map";

            // !TODO: write the robot coordinates
            geometry_msgs::msg::PoseStamped rbt_pose;
            rbt_pose.pose.position.x = 0.0;

            // !TODO: write the goal coordinates

            // !TODO: fill up the array containing the robot coordinates at [0] and goal coordinates at [1]
            msg_path_request.poses.push_back(rbt_pose);

            // Publish the message
            RCLCPP_INFO(this->get_logger(),
                        "Sending Path Planning Request from Rbt @ (%7.3f, %7.3f) to Goal @ (%7.3f, %7.3f)",
                        rbt_x_, rbt_y_, goal_x_, goal_y_);
            this->pub_path_request_->publish(msg_path_request);
        }
    };
}

// Main Boiler Plate =============================================================
int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ee3305::Behavior>());
    rclcpp::shutdown();
    return 0;
}
